src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -3679,6 +3679,74 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
+        if (!arg) {
+            return ASR::make_StringConstant_t(al, loc, "", res_type);
+        }
+        if (ASRUtils::is_real(*arg_type)) {
+            if (ASRUtils::expr_value(arg) != nullptr) {
+                double ival = ASR::down_cast<ASR::RealConstant_t>(
+                                        ASRUtils::expr_value(arg))->m_r;
+                // stringstream for exponential notation formatting.
+                std::stringstream sm;
+                sm << ival;
+                std::string value_str = sm.str();
+                sm.clear();
+                res_value =  ASR::down_cast<ASR::expr_t>(ASR::make_StringConstant_t(al,
+                                loc, s2c(al, value_str), res_type));
+            }
+            return (ASR::asr_t *)ASR::down_cast<ASR::expr_t>(ASR::make_Cast_t(
+                al, loc, arg, ASR::cast_kindType::RealToCharacter,
+                res_type, res_value));
+        } else if (ASRUtils::is_integer(*arg_type)) {
+            if (ASRUtils::expr_value(arg) != nullptr) {
+                int64_t number = ASR::down_cast<ASR::IntegerConstant_t>(
+                                        ASRUtils::expr_value(arg))->m_n;
+                std::string value_str = std::to_string(number);
+                res_value = ASR::down_cast<ASR::expr_t>(ASR::make_StringConstant_t(al,
+                                loc, s2c(al, value_str), res_type));
+            }
+            return (ASR::asr_t *)ASR::down_cast<ASR::expr_t>(ASR::make_Cast_t(
+                al, loc, arg, ASR::cast_kindType::IntegerToCharacter,
+                res_type, res_value));
+        } else if (ASRUtils::is_logical(*arg_type)) {
+            if(ASRUtils::expr_value(arg) != nullptr) {
+                bool bool_number = ASR::down_cast<ASR::LogicalConstant_t>(
+                                        ASRUtils::expr_value(arg))->m_value;
+                std::string value_str = (bool_number)? "True" : "False";
+                res_value = ASR::down_cast<ASR::expr_t>(ASR::make_StringConstant_t(al,
+                                loc, s2c(al, value_str), res_type));
+            }
+            return (ASR::asr_t *)ASR::down_cast<ASR::expr_t>(ASR::make_Cast_t(
+                al, loc, arg, ASR::cast_kindType::LogicalToCharacter,
+                res_type, res_value));
+
+        } else if (ASRUtils::is_character(*arg_type)) {
+            if(ASRUtils::expr_value(arg) != nullptr) {
+                std::string string_it_self = ASR::down_cast<ASR::StringConstant_t>(
+                                        ASRUtils::expr_value(arg))->m_s;
+                res_value = ASR::down_cast<ASR::expr_t>(ASR::make_StringConstant_t(al,
+                                loc, s2c(al, string_it_self), res_type));
+            }
+            return (ASR::asr_t *)ASR::down_cast<ASR::expr_t>(ASR::make_Cast_t(
+                al, loc, arg, ASR::cast_kindType::CharacterToCharacter,
+                res_type, res_value));
```

Comment:
It seems to me that the `CharacterToCharacter` ASR node is not needed. Can't we simply return the argument like this:
```suggestion
        } else if (ASRUtils::is_character(*arg_type)) {
            return (ASR::asr_t *)arg;
```

Reply:
it didn't work and raised the same error i posted on zulip.